solr/modules/monitor/src/java/org/apache/solr/monitor/update/MonitorUpdateRequestProcessor.java

```diff
@@ -0,0 +1,257 @@
+import org.apache.solr.core.SolrCore;
+import org.apache.solr.monitor.MonitorConstants;
+import org.apache.solr.monitor.MonitorSchemaFields;
+import org.apache.solr.monitor.SimpleQueryParser;
+import org.apache.solr.schema.IndexSchema;
+import org.apache.solr.schema.SchemaField;
+import org.apache.solr.update.AddUpdateCommand;
+import org.apache.solr.update.processor.UpdateRequestProcessor;
+
+public class MonitorUpdateRequestProcessor extends UpdateRequestProcessor {
+
+  private final SolrCore core;
+  private final IndexSchema indexSchema;
+  private final Presearcher presearcher;
+  private final MonitorSchemaFields monitorSchemaFields;
+
+  public MonitorUpdateRequestProcessor(
+      UpdateRequestProcessor next, SolrCore core, Presearcher presearcher) {
+    super(next);
+    this.core = core;
+    this.indexSchema = core.getLatestSchema();
+    this.presearcher = presearcher;
+    this.monitorSchemaFields = new MonitorSchemaFields(indexSchema);
+  }
+
+  @Override
+  public void processAdd(AddUpdateCommand cmd) throws IOException {
+    var solrInputDocument = cmd.getSolrInputDocument();
+    var queryId =
+        (String) solrInputDocument.getFieldValue(indexSchema.getUniqueKeyField().getName());
+    var queryFieldValue = solrInputDocument.getFieldValue(MonitorFields.MONITOR_QUERY);
+    if (queryFieldValue != null) {
+      var payload =
+          Optional.ofNullable(solrInputDocument.getFieldValue(MonitorFields.PAYLOAD))
+              .map(Object::toString)
+              .orElse(null);
+      List<SolrInputDocument> children =
+          Optional.of(queryFieldValue)
+              .filter(String.class::isInstance)
+              .map(String.class::cast)
+              .map(
+                  queryStr ->
+                      new MonitorQuery(
+                          queryId, SimpleQueryParser.parse(queryStr, core), queryStr, Map.of()))
+              .stream()
+              .flatMap(monitorQuery -> decompose(monitorQuery, payload))
+              .map(this::toSolrInputDoc)
+              .collect(Collectors.toList());
+      if (children.isEmpty()) {
+        throw new SolrException(
+            SolrException.ErrorCode.INVALID_STATE, "Query could not be decomposed");
+      }
+      SolrInputDocument firstChild = children.get(0);
+      if (solrInputDocument.hasChildDocuments()) {
+        solrInputDocument.getChildDocuments().clear();
+      }
+      solrInputDocument.addChildDocuments(children.stream().skip(1).collect(Collectors.toList()));
+      if (solrInputDocument.hasChildDocuments()) {
+        solrInputDocument
+            .getChildDocuments()
+            .forEach(
+                child ->
+                    solrInputDocument.forEach(
+                        field -> {
+                          if (!MonitorFields.RESERVED_MONITOR_FIELDS.contains(field.getName())) {
+                            child.addField(field.getName(), field.getValue());
+                          }
```

Comment:
Oh and from code reading in this class so far here I had a question here: so this is copying-over non-reserved fields, right? Beyond the query id and query payload and query meta data, when would there be what additional fields in a typical use case?

Reply:
The idea is that a parent query "doc" can get decomposed into many child "docs" _without_ the caller knowing it. If they send the query "doc" with some additional non-monitor data fields stored alongside in separate columns then in order to guarantee that the column data will be returned on _any_ match you have to copy over that extra data to all the child "docs" (most likely query disjuncts).

Reply:
> ... If they send the query "doc" with some additional non-monitor data fields stored alongside in separate columns ...

Naive question: could we just disallow that (i.e. reject documents that contain additional fields) generally or for the initial phase if there's a phased approach? Or are additional non-monitor data fields needed even for a basic typical use case?

Reply:
If you feel strongly about this I can reconsider it but because supporting it feels relatively easy from an implementation perspective, I am more inclined to keeping it. Put another way, what is your objection to this specifically, is it the "hidden" (aka opaque to the user) denormalization or is it the implementation complexity?

Reply:
I see a mix of 'engineering' and 'product' complexity here, perhaps relatively marginal engineering complexity but still, and it's not yet clear to me when/why additional non-monitor data fields would be sent from a use case/product perspective.

To elaborate a bit more:
* 'engineering' aspects: 
  * supporting additional non-monitor data fields introduces the 'reserved vs. non-reserved monitor fields' concept i.e.
  * the non-reserved fields need to be copied over but the reserved fields must not be copied over because they are already there.
  * this needs to be understood at the implementation level and covered with test functionality.
  * the concept is present in the update processor and in the aliasing prefetcher.
* 'product' aspects:
  * _"pass (only) a query id and a query in each document"_ is a very accessible user interface.
  * _"pass (only) a query id and a query in each document, and optionally a payload"_ is a more sophisticated user interface and more explanation/documentation is needed about the optional payload.
  * _"pass a query id and a query in each document, and optionally a payload, and anything else you like with some caveats"_ seems a less accessible user interface, with more choice i.e. decision making required of the user.

My preference for an initial phase would be the minimal product approach i.e. disallowing additional fields.

Reply:
A concrete use-case might be storing a user id along with the query to know which user to forward the alert to. So if I'm understanding correctly you recommend removing both the ability to write to additional columns as well as the payload field? We'd probably have to introduce some validation making sure the "document" contained no other fields other than the query.

Reply:
Sorry for the delay in replying here.

> ... So if I'm understanding correctly you recommend removing both the ability to write to additional columns as well as the payload field? ...

Yes, I think a phased approach and an initially minimum basic solution would make the integration more approachable and accessible from both a reviewing and users-getting-started perspective. But maybe my preference and perspective is unusual here, more evolution than revolution, or something. Perhaps others reading this here on this pull request would like to chime in on this and/or something to explore on the https://lists.apache.org/thread/yp47w0mjqblyxvrpd5rbpd4qh9m9o7qq dev mailing list thread maybe?

> ... We'd probably have to introduce some validation making sure the "document" contained no other fields other than the query.

Yes, validation for the presence and only the presence of all the required fields.

And as an aside the `if (queryFieldValue != null) {` check puzzled me a little, well, not the check but the `else` branch being absent and not an error, though perhaps I'm still missing some subtlety there w.r.t. decomposition/decoding/persistence/etc. of the documents.

Reply:
I've removed the payload and other field support because I don't want to overload that thread. There is probably enough to discuss otherwise. I've also added validations which cover, among other things, the puzzling `queryFieldValue == null` condition. I think I didn't validate this before because until now my tests never specified custom update processor chains but I also had some tests which ran regular, non-monitor-query updates.